File: scrape_angebote.py
"""
MealDeal Angebote-Scraper
=========================
Holt aktuelle Supermarkt-Angebote von Marktguru
und speichert sie in der Supabase-Datenbank.

Nutzung: python scrape_angebote.py
        oder Doppelklick auf ANGEBOTE_LADEN.bat
"""

import json
import re
import time
import sys
import os

# === Pakete pruefen ===
try:
    import requests
except ImportError:
    print("Installiere benoetigte Pakete...")
    os.system(f"{sys.executable} -m pip install requests --break-system-packages -q")
    import requests

# === SUPABASE CONFIG ===
SUPABASE_URL = "https://wjhesvkapqrsbibqjbtr.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6IndqaGVzdmthcHFyc2JpYnFqYnRyIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzI4NTE4NDEsImV4cCI6MjA4ODQyNzg0MX0.-Vh6_Qtz1EZzPJA70lXf8boCGBs1S7c05KzbUFK-dJs"

# === MARKTGURU CONFIG ===
MARKTGURU_SITE = "https://www.marktguru.de"
MARKTGURU_API = "https://api.marktguru.de/api/v1"

# Supermaerkte
KNOWN_MARKETS = {
    "REWE": "REWE", "REWE Center": "REWE",
    "PENNY": "Penny",
    "EDEKA": "Edeka", "E center": "Edeka",
    "Kaufland": "Kaufland",
    "Lidl": "Lidl",
    "ALDI SÜD": "ALDI", "ALDI Nord": "ALDI",
    "Netto Marken-Discount": "Netto", "Netto": "Netto",
    "Norma": "Norma",
    "nahkauf": "REWE",
}

# Obst-Keywords fuer Obst/Gemuese-Unterscheidung
OBST = ["apfel", "banane", "orange", "birne", "kirsche", "erdbeere", "beere",
        "traube", "mango", "ananas", "melone", "zitrone", "limette", "pflaume",
        "pfirsich", "nektarine", "kiwi", "obst", "clementine", "mandarine",
        "himbeere", "heidelbeere", "weintraube", "grapefruit"]


def map_category(cat_name, product_name):
    """Mapped Marktguru-Kategorie auf unsere Kategorien"""
    cat = (cat_name or "").lower()
    name = product_name.lower()

    if any(k in cat for k in ["fleisch", "wurst", "schinken", "geflügel", "rind"]):
        return "Fleisch"
    if any(k in cat for k in ["obst", "gemüse", "salat"]):
        return "Obst" if any(k in name for k in OBST) else "Gemüse"
    if any(k in cat for k in ["milch", "joghurt", "butter", "sahne", "quark", "eier"]):
        return "Milch & Eier"
    if "käse" in cat:
        return "Käse"
    if "tiefkühl" in cat or "tk-" in cat:
        return "Tiefkühl"
    if any(k in cat for k in ["getränke", "bier", "wein", "saft", "wasser"]):
        return "Getränke"
    if any(k in cat for k in ["snack", "süß", "chips", "schokolade", "keks"]):
        return "Snacks & Süßes"
    if any(k in cat for k in ["brot", "back", "brötchen"]):
        return "Backwaren"
    if any(k in cat for k in ["nudel", "reis", "pasta"]):
        return "Nudeln & Reis"
    if any(k in cat for k in ["konserve", "fertig", "dose"]):
        return "Konserven"
    if any(k in cat for k in ["gewürz", "sauce", "soße", "essig"]):
        return "Gewürze"
    if any(k in cat for k in ["fisch", "meeresfrüchte"]):
        return "Fisch & Meeresfrüchte"
    if any(k in cat for k in ["öl", "fett"]):
        return "Öle & Fette"
    if any(k in cat for k in ["haushalt", "reinigung"]):
        return "Haushalt"
    if any(k in cat for k in ["drogerie", "pflege", "hygiene"]):
        return "Drogerie"
    return "Sonstiges Lebensmittel"


def extract_api_keys():
    """Extrahiert API-Keys von der Marktguru-Website"""
    print("Hole API-Keys von marktguru.de...", end=" ")
    try:
        resp = requests.get(MARKTGURU_SITE, timeout=15, headers={
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0"
        })
        resp.raise_for_status()

        api_key = re.search(r'"apiKey":"([^"]+)"', resp.text)
        client_key = re.search(r'"clientKey":"([^"]+)"', resp.text)

        if api_key and client_key:
            print("OK")
            return api_key.group(1), client_key.group(1)
        else:
            print("Keys nicht gefunden im HTML")
            return None, None
    except Exception as e:
        print(f"Fehler: {e}")
        return None, None


def fetch_offers(plz, api_key, client_key, max_offers=1500):
    """Holt Angebote von der Marktguru API"""
    session = requests.Session()
    session.headers.update({
        "Accept": "application/json",
        "x-apikey": api_key,
        "x-clientkey": client_key,
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0",
    })

    all_offers = []
    BATCH = 200
    industries = [1009, 1023]  # Supermarkt + Discounter

    for industry_id in industries:
        industry_name = "Supermarkt" if industry_id == 1009 else "Discounter"
        print(f"\n  [{industry_name}] Lade Angebote...")
        offset = 0

        while offset < max_offers:
            params = {
                "as": "web",
                "limit": BATCH,
                "offset": offset,
                "zipCode": plz,
                "industryId": industry_id,
            }

            try:
                time.sleep(0.5)  # Rate limit
                resp = session.get(f"{MARKTGURU_API}/offers", params=params, timeout=20)
                resp.raise_for_status()
                data = resp.json()

                # API Format parsen
                if isinstance(data, list):
                    offers = data
                elif isinstance(data, dict):
                    offers = data.get("results", data.get("offers", data.get("data", [])))
                else:
                    break

                if not offers:
                    break

                all_offers.extend(offers)
                print(f"    +{len(offers)} Angebote (gesamt: {len(all_offers)})")

                if len(offers) < BATCH:
                    break
                offset += BATCH

            except requests.exceptions.HTTPError as e:
                print(f"    HTTP-Fehler: {e.response.status_code}")
                break
            except Exception as e:
                print(f"    Fehler: {e}")
                break

    return all_offers


def save_to_supabase(offers_data):
    """Speichert Angebote in Supabase"""
    if not offers_data:
        return 0

    url = f"{SUPABASE_URL}/rest/v1/offers"
    headers = {
        "apikey": SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "application/json",
        "Prefer": "resolution=merge-duplicates",
    }

    total = 0
    BATCH = 100

    for i in range(0, len(offers_data), BATCH):
        batch = offers_data[i:i+BATCH]
        try:
            resp = requests.post(url, json=batch, headers=headers, timeout=30)
            if resp.status_code in [200, 201]:
                total += len(batch)
                print(f"  Batch {i//BATCH + 1}: {len(batch)} gespeichert")
            else:
                print(f"  Batch {i//BATCH + 1} Fehler ({resp.status_code}): {resp.text[:150]}")
        except Exception as e:
            print(f"  Batch {i//BATCH + 1} Fehler: {e}")

    return total


def main():
    print("=" * 50)
    print("  MealDeal - Angebote laden")
    print("=" * 50)
    print()

    # PLZ abfragen
    plz = input("Deine PLZ eingeben: ").strip()
    if len(plz) != 5 or not plz.isdigit():
        print("Ungueltige PLZ! Muss 5 Ziffern sein.")
        return

    plz_prefix = plz[:3]

    # 1. API-Keys holen
    api_key, client_key = extract_api_keys()
    if not api_key or not client_key:
        print("\nKonnte keine API-Keys bekommen. Bitte spaeter nochmal versuchen.")
        return

    # 2. Angebote von Marktguru holen
    print(f"\nHole Angebote fuer PLZ {plz}...")
    raw_offers = fetch_offers(plz, api_key, client_key)
    print(f"\n{len(raw_offers)} Angebote von Marktguru erhalten")

    if not raw_offers:
        print("Keine Angebote gefunden.")
        return

    # 3. Angebote transformieren
    transformed = []
    store_counts = {}

    for offer in raw_offers:
        advertiser = offer.get("advertiserName", "").strip()
        market = KNOWN_MARKETS.get(advertiser)
        if not market:
            for key, val in KNOWN_MARKETS.items():
                if key.upper() in advertiser.upper():
                    market = val
                    break
        if not market:
            continue

        title = offer.get("title", "").strip()
        if not title:
            continue

        price = offer.get("price")
        if price is None:
            continue

        ref_price = offer.get("referencePrice")
        discount = None
        if ref_price and ref_price > 0 and ref_price > price:
            discount = round(((ref_price - price) / ref_price) * 100)

        # Kategorie
        cat_name = ""
        cat_obj = offer.get("category")
        if isinstance(cat_obj, dict):
            cat_name = cat_obj.get("name", "")
        elif isinstance(cat_obj, str):
            cat_name = cat_obj

        category = map_category(cat_name, title)

        valid_from = (offer.get("validFrom") or "")[:10] or None
        valid_until = (offer.get("validTo") or "")[:10]
        if not valid_until:
            continue

        fingerprint = f"{market}_{title}_{valid_until}".lower().strip()

        record = {
            "product_name": title,
            "store": market,
            "offer_price": round(float(price), 2),
            "original_price": round(float(ref_price), 2) if ref_price else None,
            "discount_percent": discount,
            "plz": plz,
            "plz_prefix": plz_prefix,
            "category": category,
            "valid_from": valid_from,
            "valid_until": valid_until,
            "image_url": offer.get("imageUrl"),
            "quantity": offer.get("quantity"),
            "fingerprint": fingerprint,
        }

        transformed.append(record)
        store_counts[market] = store_counts.get(market, 0) + 1

    print(f"\n{len(transformed)} Angebote nach Filterung:\n")
    for store, count in sorted(store_counts.items(), key=lambda x: -x[1]):
        print(f"  {store}: {count} Angebote")

    if not transformed:
        print("\nKeine Angebote zum Speichern.")
        return

    # 4. In Supabase speichern
    print(f"\nSpeichere {len(transformed)} Angebote in Supabase...")
    saved = save_to_supabase(transformed)

    print(f"\n{'=' * 50}")
    print(f"  FERTIG! {saved} Angebote gespeichert")
    print(f"{'=' * 50}")
    print(f"\n  Starte jetzt die App mit STARTE_APP.bat")
    print(f"  um die Angebote zu sehen!")


if __name__ == "__main__":
    main()
    input("\nDruecke Enter zum Schliessen...")

File: _archiv/alte-scraper/scrape_angebote.js
/**
 * MealDeal - Angebote Scraper (Node.js)
 * Holt Angebote von Marktguru und speichert sie in Supabase
 */

const https = require("https");
const readline = require("readline");

// === CONFIG ===
const SUPABASE_URL = "https://wjhesvkapqrsbibqjbtr.supabase.co";
const SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6IndqaGVzdmthcHFyc2JpYnFqYnRyIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzI4NTE4NDEsImV4cCI6MjA4ODQyNzg0MX0.-Vh6_Qtz1EZzPJA70lXf8boCGBs1S7c05KzbUFK-dJs";

const KNOWN_MARKETS = {
  "REWE": "REWE", "REWE Center": "REWE",
  "PENNY": "Penny",
  "EDEKA": "Edeka", "E center": "Edeka",
  "Kaufland": "Kaufland",
  "Lidl": "Lidl",
  "ALDI SÜD": "ALDI", "ALDI Nord": "ALDI",
  "Netto Marken-Discount": "Netto", "Netto": "Netto",
  "Norma": "Norma", "nahkauf": "REWE",
};

const OBST = ["apfel","banane","orange","birne","kirsche","erdbeere","beere","traube","mango","ananas","melone","zitrone","pflaume","pfirsich","nektarine","kiwi","obst","clementine","mandarine","himbeere","heidelbeere"];

function mapCategory(catName, productName) {
  const cat = (catName || "").toLowerCase();
  const name = productName.toLowerCase();
  if (["fleisch","wurst","schinken","geflügel"].some(k => cat.includes(k))) return "Fleisch";
  if (["obst","gemüse","salat"].some(k => cat.includes(k))) return OBST.some(k => name.includes(k)) ? "Obst" : "Gemüse";
  if (["milch","joghurt","butter","sahne","quark","eier"].some(k => cat.includes(k))) return "Milch & Eier";
  if (cat.includes("käse")) return "Käse";
  if (cat.includes("tiefkühl")) return "Tiefkühl";
  if (["getränke","bier","wein","saft","wasser"].some(k => cat.includes(k))) return "Getränke";
  if (["snack","süß","chips","schokolade"].some(k => cat.includes(k))) return "Snacks & Süßes";
  if (["brot","back","brötchen"].some(k => cat.includes(k))) return "Backwaren";
  if (["nudel","reis","pasta"].some(k => cat.includes(k))) return "Nudeln & Reis";
  if (["konserve","fertig"].some(k => cat.includes(k))) return "Konserven";
  if (["gewürz","sauce","soße","essig"].some(k => cat.includes(k))) return "Gewürze";
  if (["fisch","meeresfrüchte"].some(k => cat.includes(k))) return "Fisch & Meeresfrüchte";
  if (["öl","fett"].some(k => cat.includes(k))) return "Öle & Fette";
  if (["haushalt","reinigung"].some(k => cat.includes(k))) return "Haushalt";
  if (["drogerie","pflege","hygiene"].some(k => cat.includes(k))) return "Drogerie";
  return "Sonstiges Lebensmittel";
}

// HTTPS GET helper
function httpsGet(url, headers = {}) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, { headers }, (res) => {
      let data = "";
      res.on("data", (chunk) => data += chunk);
      res.on("end", () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try { resolve(JSON.parse(data)); } catch { resolve(data); }
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data.slice(0, 200)}`));
        }
      });
    });
    req.on("error", reject);
    req.setTimeout(20000, () => { req.destroy(); reject(new Error("Timeout")); });
  });
}

// HTTPS POST helper
function httpsPost(url, body, headers = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const postData = JSON.stringify(body);
    const options = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      method: "POST",
      headers: { ...headers, "Content-Type": "application/json", "Content-Length": Buffer.byteLength(postData) },
    };
    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (chunk) => data += chunk);
      res.on("end", () => {
        if (res.statusCode >= 200 && res.statusCode < 300) resolve(data);
        else reject(new Error(`HTTP ${res.statusCode}: ${data.slice(0, 200)}`));
      });
    });
    req.on("error", reject);
    req.setTimeout(30000, () => { req.destroy(); reject(new Error("Timeout")); });
    req.write(postData);
    req.end();
  });
}

// Marktguru API-Keys extrahieren
async function getApiKeys() {
  process.stdout.write("Hole API-Keys von marktguru.de... ");
  try {
    const html = await new Promise((resolve, reject) => {
      https.get("https://www.marktguru.de", {
        headers: { "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0" }
      }, (res) => {
        let data = "";
        res.on("data", (chunk) => data += chunk);
        res.on("end", () => resolve(data));
      }).on("error", reject);
    });

    const apiKey = html.match(/"apiKey":"([^"]+)"/);
    const clientKey = html.match(/"clientKey":"([^"]+)"/);

    if (apiKey && clientKey) {
      console.log("OK");
      return { apiKey: apiKey[1], clientKey: clientKey[1] };
    }
    console.log("Keys nicht im HTML gefunden");
    return null;
  } catch (e) {
    console.log("Fehler: " + e.message);
    return null;
  }
}

// Angebote von Marktguru holen
async function fetchOffers(plz, apiKey, clientKey) {
  const allOffers = [];
  const industries = [1009, 1023]; // Supermarkt + Discounter

  for (const industryId of industries) {
    const label = industryId === 1009 ? "Supermarkt" : "Discounter";
    console.log(`\n  [${label}] Lade Angebote...`);
    let offset = 0;

    while (offset < 1500) {
      const url = `https://api.marktguru.de/api/v1/offers?as=web&limit=200&offset=${offset}&zipCode=${plz}&industryId=${industryId}`;
      try {
        await new Promise(r => setTimeout(r, 500)); // Rate limit
        const data = await httpsGet(url, {
          "Accept": "application/json",
          "x-apikey": apiKey,
          "x-clientkey": clientKey,
          "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0",
        });

        const offers = Array.isArray(data) ? data : (data?.results || data?.offers || data?.data || []);
        if (!offers.length) break;

        allOffers.push(...offers);
        console.log(`    +${offers.length} Angebote (gesamt: ${allOffers.length})`);
        if (offers.length < 200) break;
        offset += 200;
      } catch (e) {
        console.log(`    Fehler: ${e.message}`);
        break;
      }
    }
  }
  return allOffers;
}

// In Supabase speichern
async function saveToSupabase(records) {
  if (!records.length) return 0;
  let total = 0;
  const BATCH = 100;

  for (let i = 0; i < records.length; i += BATCH) {
    const batch = records.slice(i, i + BATCH);
    try {
      await httpsPost(`${SUPABASE_URL}/rest/v1/offers`, batch, {
        "apikey": SUPABASE_KEY,
        "Authorization": `Bearer ${SUPABASE_KEY}`,
        "Prefer": "resolution=merge-duplicates",
      });
      total += batch.length;
      console.log(`  Batch ${Math.floor(i/BATCH)+1}: ${batch.length} gespeichert`);
    } catch (e) {
      console.log(`  Batch ${Math.floor(i/BATCH)+1} Fehler: ${e.message}`);
    }
  }
  return total;
}

// Angebote transformieren
function transformOffers(rawOffers, plz) {
  const plzPrefix = plz.slice(0, 3);
  const transformed = [];
  const storeCounts = {};

  for (const offer of rawOffers) {
    const advertiser = (offer.advertiserName || "").trim();
    let market = KNOWN_MARKETS[advertiser];
    if (!market) {
      for (const [key, val] of Object.entries(KNOWN_MARKETS)) {
        if (advertiser.toUpperCase().includes(key.toUpperCase())) { market = val; break; }
      }
    }
    if (!market) continue;

    const title = (offer.title || "").trim();
    if (!title) continue;
    const price = offer.price;
    if (price == null) continue;

    const refPrice = offer.referencePrice;
    let discount = null;
    if (refPrice && refPrice > 0 && refPrice > price) {
      discount = Math.round(((refPrice - price) / refPrice) * 100);
    }

    let catName = "";
    if (offer.category && typeof offer.category === "object") catName = offer.category.name || "";
    else if (typeof offer.category === "string") catName = offer.category;

    const validUntil = (offer.validTo || "").slice(0, 10);
    if (!validUntil) continue;

    transformed.push({
      product_name: title,
      store: market,
      offer_price: Math.round(price * 100) / 100,
      original_price: refPrice ? Math.round(refPrice * 100) / 100 : null,
      discount_percent: discount,
      plz: plz,
      plz_prefix: plzPrefix,
      category: mapCategory(catName, title),
      valid_from: (offer.validFrom || "").slice(0, 10) || null,
      valid_until: validUntil,
      image_url: offer.imageUrl || null,
      quantity: offer.quantity || null,
      fingerprint: `${market}_${title}_${validUntil}`.toLowerCase().trim(),
    });

    storeCounts[market] = (storeCounts[market] || 0) + 1;
  }

  return { transformed, storeCounts };
}

// === MAIN ===
async function main() {
  console.log("=".repeat(50));
  console.log("  MealDeal - Angebote laden");
  console.log("=".repeat(50));
  console.log();

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const plz = await new Promise(r => rl.question("Deine PLZ eingeben: ", r));
  rl.close();

  if (!/^\d{5}$/.test(plz.trim())) {
    console.log("Ungueltige PLZ! Muss 5 Ziffern sein.");
    return;
  }

  // 1. API-Keys
  const keys = await getApiKeys();
  if (!keys) { console.log("\nAPI-Keys nicht verfuegbar."); return; }

  // 2. Angebote holen
  console.log(`\nHole Angebote fuer PLZ ${plz.trim()}...`);
  const raw = await fetchOffers(plz.trim(), keys.apiKey, keys.clientKey);
  console.log(`\n${raw.length} Angebote von Marktguru erhalten`);
  if (!raw.length) { console.log("Keine Angebote gefunden."); return; }

  // 3. Transformieren
  const { transformed, storeCounts } = transformOffers(raw, plz.trim());
  console.log(`\n${transformed.length} Angebote nach Filterung:\n`);
  Object.entries(storeCounts).sort((a,b) => b[1]-a[1]).forEach(([s,c]) => console.log(`  ${s}: ${c} Angebote`));

  if (!transformed.length) { console.log("\nKeine Angebote zum Speichern."); return; }

  // 4. Speichern
  console.log(`\nSpeichere ${transformed.length} Angebote in Supabase...`);
  const saved = await saveToSupabase(transformed);

  console.log(`\n${"=".repeat(50)}`);
  console.log(`  FERTIG! ${saved} Angebote gespeichert`);
  console.log("=".repeat(50));
  console.log(`\n  Starte jetzt die App mit STARTE_APP.bat`);
}

main().catch(e => console.error("Fehler:", e.message));
